tiledb/sm/filter/bitsort_filter.cc

```diff
@@ -0,0 +1,465 @@
+  // Output size does not change with this filter.
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Write the metadata.
+  auto parts = input->buffers();
+  auto num_parts = (uint32_t)parts.size();
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // Keep track of the starting offsets to write in positions for each part.
+  std::vector<uint32_t> offsets;
+  uint32_t total_size = 0;
+  for (const auto& part : parts) {
+    auto part_size = (uint32_t)part.size();
+    offsets.push_back(total_size);
+    total_size += part_size;
+  }
+
+  // Sort all parts.
+  std::vector<std::pair<T, uint64_t>> sorted_elements(total_size / sizeof(T));
+  for (uint64_t i = 0; i < num_parts; ++i) {
+    const auto &part = parts[i];
+    auto part_size = static_cast<uint32_t>(part.size());
+    RETURN_NOT_OK(output_metadata->write(&part_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(sort_part<T>(&part, output_buf, offsets[i], sorted_elements));
+  }
+
+  // Rewrite each of the dimension tile data with the new sort order.
+  std::vector<Tile*> &dim_tiles = pair.first.get();
+
+  // Since rewrite_dim_tile_forward is only moving around dimension data and 
+  // reordering it, the type becomes irrelevant and only the size of the type
+  // matters, so the code cases on the size of the type and passes in the
+  // corresponding integer type into a templated function.
+  for (auto* dim_tile : dim_tiles) {
+    Datatype tile_type = dim_tile->type();
+    switch (datatype_size(tile_type)) {
+      case sizeof(uint8_t): {
+        rewrite_dim_tile_forward<T, uint8_t>(sorted_elements, dim_tile);
+      } break;
+      case sizeof(uint16_t): {
+        rewrite_dim_tile_forward<T, uint16_t>(sorted_elements, dim_tile);
+      } break;
+      case sizeof(uint32_t): {
+        rewrite_dim_tile_forward<T, uint32_t>(sorted_elements, dim_tile);
+      } break;
+      case sizeof(uint64_t): {
+        rewrite_dim_tile_forward<T, uint64_t>(sorted_elements, dim_tile);
+      } break;
+      default: {
+        return Status_FilterError(
+          "BitSortFilter::sort_part: dimension datatype does not have an appropriate "
+          "size");
+      }
+    }
+  }
+
+  return Status::Ok();
+}
+
+template <typename T>
+Status BitSortFilter::sort_part(const ConstBuffer* input_buffer, Buffer* output_buffer, uint32_t start, std::vector<std::pair<T, uint64_t>> &sorted_elements) const {
+  // Read in the data.
+  uint32_t s = input_buffer->size();
+  assert(s % sizeof(T) == 0);
+  uint32_t num_elems_in_part = s / sizeof(T);
+
+  if (num_elems_in_part == 0) {
+    return Status::Ok();
+  }
+  const T* part_array = static_cast<const T*>(input_buffer->data());
```

Comment:
I'm not sure what part_array is? A comment above or a better name might help clarify.

Reply:
It is just the input data in pointer form? I changed the name to input_data if that helps